File: src/shared/logging/telemetry.ts
import {
  debug as pluginDebug,
  error as pluginError,
  info as pluginInfo,
  warn as pluginWarn,
} from "@tauri-apps/plugin-log";

export type TelemetryLevel = "debug" | "info" | "warn" | "error";

type TelemetryPayload = {
  event: string;
  level: TelemetryLevel;
  traceId: string | null;
} & Record<string, unknown>;

/** 生成 traceId。 */
export function createTraceId() {
  if (
    typeof crypto !== "undefined" &&
    typeof crypto.randomUUID === "function"
  ) {
    return crypto.randomUUID();
  }
  return `${Date.now()}-${Math.random().toString(16).slice(2)}`;
}

/** 输出统一结构化埋点。 */
export async function logTelemetry(
  level: TelemetryLevel,
  event: string,
  fields?: Record<string, unknown>,
) {
  const normalizedEvent = normalizeEventName(event);
  const payload: TelemetryPayload = {
    event: normalizedEvent,
    ts: Date.now(),
    source: "frontend",
    level,
    traceId:
      typeof fields?.traceId === "string" ? (fields.traceId as string) : null,
    ...(fields ?? {}),
  };
  const line = JSON.stringify(payload);
  if (level === "debug") return pluginDebug(line);
  if (level === "warn") return pluginWarn(line);
  if (level === "error") return pluginError(line);
  return pluginInfo(line);
}

function normalizeEventName(value: string): string {
  const withDots = value
    .trim()
    .replace(/[:_]/g, ".")
    .replace(/([a-z0-9])([A-Z])/g, "$1.$2")
    .replace(/\.+/g, ".")
    .replace(/^\./, "")
    .replace(/\.$/, "");
  return withDots.toLowerCase() || "frontend.log.update";
}

function parseTelemetryLikeMessage(input: string): {
  event: string;
  fields?: Record<string, unknown>;
} {
  const trimmed = input.trim();
  if (!trimmed) {
    return {
      event: "frontend.log.update",
      fields: { message: "" },
    };
  }
  try {
    const parsed = JSON.parse(trimmed) as Record<string, unknown>;
    if (parsed && typeof parsed === "object") {
      const event =
        typeof parsed.event === "string" && parsed.event.trim()
          ? parsed.event
          : "frontend.log.update";
      const { event: _event, ...rest } = parsed;
      return {
        event,
        fields: rest,
      };
    }
  } catch {
    // 非 JSON 字符串按 message 透传。
  }
  return {
    event: "frontend.log.update",
    fields: { message: input },
  };
}

async function logRaw(level: TelemetryLevel, message: string) {
  const { event, fields } = parseTelemetryLikeMessage(message);
  return logTelemetry(level, event, fields);
}

export async function debug(message: string) {
  return logRaw("debug", message);
}

export async function info(message: string) {
  return logRaw("info", message);
}

export async function warn(message: string) {
  return logRaw("warn", message);
}

export async function error(message: string) {
  return logRaw("error", message);
}

File: crates/engine/src/telemetry.rs
//! 统一埋点日志助手。
use std::time::{SystemTime, UNIX_EPOCH};

use log::{debug, error, info, warn};
use serde_json::{Map, Value, json};

/// 埋点级别。
pub enum TelemetryLevel {
    Debug,
    Info,
    Warn,
    Error,
}

/// 输出统一结构化埋点日志。
pub fn log_telemetry(level: TelemetryLevel, event: &str, trace_id: Option<&str>, fields: Value) {
    let mut payload = Map::new();
    payload.insert("event".to_string(), Value::String(event.to_string()));
    payload.insert("ts".to_string(), json!(now_epoch_ms()));
    payload.insert("source".to_string(), Value::String("engine".to_string()));
    payload.insert(
        "level".to_string(),
        Value::String(
            match level {
                TelemetryLevel::Debug => "debug",
                TelemetryLevel::Info => "info",
                TelemetryLevel::Warn => "warn",
                TelemetryLevel::Error => "error",
            }
            .to_string(),
        ),
    );
    payload.insert(
        "traceId".to_string(),
        trace_id
            .map(|v| Value::String(v.to_string()))
            .unwrap_or(Value::Null),
    );
    if let Value::Object(map) = fields {
        map.into_iter().for_each(|(k, v)| {
            payload.insert(k, v);
        });
    }
    let line = Value::Object(payload).to_string();
    match level {
        TelemetryLevel::Debug => debug!("{line}"),
        TelemetryLevel::Info => info!("{line}"),
        TelemetryLevel::Warn => warn!("{line}"),
        TelemetryLevel::Error => error!("{line}"),
    }
}

fn now_epoch_ms() -> u128 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis())
        .unwrap_or(0)
}
